File: tests_python/test_adaboost.py
#!/usr/bin/env python3
"""
Test Suite for AuroraML AdaBoost Algorithms
Tests AdaBoostClassifier and AdaBoostRegressor
"""

import sys
import os
import unittest
import numpy as np

# Add the build directory to Python path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'build'))

class TestAdaBoostClassifier(unittest.TestCase):
    """Test AdaBoostClassifier algorithm"""
    
    def setUp(self):
        """Set up test data"""
        np.random.seed(42)
        self.n_samples = 200
        self.n_features = 4
        
        # Create classification data
        self.X = np.random.randn(self.n_samples, self.n_features).astype(np.float64)
        self.y = ((self.X[:, 0] + self.X[:, 1] - 0.5 * self.X[:, 2]) > 0).astype(np.int32)
        self.X_test = np.random.randn(30, self.n_features).astype(np.float64)
        
    def test_basic_functionality(self):
        """Test basic fit and predict functionality"""
        import auroraml.adaboost as aml_adaboost
        
        model = aml_adaboost.AdaBoostClassifier(n_estimators=50, learning_rate=1.0, random_state=42)
        model.fit(self.X, self.y)
        predictions = model.predict(self.X_test)
        
        self.assertEqual(len(predictions), len(self.X_test))
        self.assertIsInstance(predictions, np.ndarray)
        
        # Check predictions are valid classes
        classes = model.classes()
        self.assertTrue(np.all(np.isin(predictions, classes)))
        
    def test_predict_proba(self):
        """Test probability prediction"""
        import auroraml.adaboost as aml_adaboost
        
        model = aml_adaboost.AdaBoostClassifier(n_estimators=50, learning_rate=1.0, random_state=42)
        model.fit(self.X, self.y)
        probabilities = model.predict_proba(self.X_test)
        
        self.assertEqual(probabilities.shape, (len(self.X_test), 2))
        self.assertTrue(np.allclose(np.sum(probabilities, axis=1), 1.0, atol=1e-5))
        self.assertTrue(np.all(probabilities >= 0))
        self.assertTrue(np.all(probabilities <= 1))
        
    def test_decision_function(self):
        """Test decision function"""
        import auroraml.adaboost as aml_adaboost
        
        model = aml_adaboost.AdaBoostClassifier(n_estimators=50, learning_rate=1.0, random_state=42)
        model.fit(self.X, self.y)
        decision = model.decision_function(self.X_test)
        
        self.assertEqual(len(decision), len(self.X_test))
        self.assertIsInstance(decision, np.ndarray)
        
    def test_parameters(self):
        """Test parameter getter and setter"""
        import auroraml.adaboost as aml_adaboost
        
        model = aml_adaboost.AdaBoostClassifier(n_estimators=50, learning_rate=1.0, random_state=42)
        
        params = model.get_params()
        self.assertIn('n_estimators', params)
        self.assertIn('learning_rate', params)
        self.assertIn('random_state', params)
        
        # Test parameter setting
        model.set_params(n_estimators=75, learning_rate=0.8)
        updated_params = model.get_params()
        self.assertEqual(updated_params['n_estimators'], "75")
        self.assertEqual(updated_params['learning_rate'], "0.800000")
        
    def test_performance(self):
        """Test model performance"""
        import auroraml.adaboost as aml_adaboost
        import auroraml.metrics as aml_metrics
        
        model = aml_adaboost.AdaBoostClassifier(n_estimators=100, learning_rate=1.0, random_state=42)
        model.fit(self.X, self.y)
        predictions = model.predict(self.X)
        
        accuracy = aml_metrics.accuracy_score(self.y, predictions)
        self.assertGreater(accuracy, 0.7)
        
    def test_different_learning_rates(self):
        """Test with different learning rates"""
        import auroraml.adaboost as aml_adaboost
        
        learning_rates = [0.5, 1.0, 2.0]
        for lr in learning_rates:
            model = aml_adaboost.AdaBoostClassifier(n_estimators=50, learning_rate=lr, random_state=42)
            model.fit(self.X, self.y)
            predictions = model.predict(self.X_test)
            self.assertEqual(len(predictions), len(self.X_test))
            
    def test_different_n_estimators(self):
        """Test with different numbers of estimators"""
        import auroraml.adaboost as aml_adaboost
        
        n_estimators_list = [10, 50, 100]
        for n_est in n_estimators_list:
            model = aml_adaboost.AdaBoostClassifier(n_estimators=n_est, learning_rate=1.0, random_state=42)
            model.fit(self.X, self.y)
            predictions = model.predict(self.X_test)
            self.assertEqual(len(predictions), len(self.X_test))
            
    def test_is_fitted(self):
        """Test is_fitted method"""
        import auroraml.adaboost as aml_adaboost
        
        model = aml_adaboost.AdaBoostClassifier()
        self.assertFalse(model.is_fitted())
        
        model.fit(self.X, self.y)
        self.assertTrue(model.is_fitted())
        
    def test_classes_attribute(self):
        """Test classes attribute"""
        import auroraml.adaboost as aml_adaboost
        
        model = aml_adaboost.AdaBoostClassifier(n_estimators=50, random_state=42)
        model.fit(self.X, self.y)
        
        classes = model.classes()
        self.assertGreaterEqual(len(classes), 2)
        self.assertTrue(np.all(np.isin(classes, [0, 1])))

class TestAdaBoostRegressor(unittest.TestCase):
    """Test AdaBoostRegressor algorithm"""
    
    def setUp(self):
        """Set up test data"""
        np.random.seed(42)
        self.n_samples = 200
        self.n_features = 4
        
        # Create regression data
        self.X = np.random.randn(self.n_samples, self.n_features).astype(np.float64)
        self.y = (2.0 * self.X[:, 0] + 1.5 * self.X[:, 1] - 
                  0.8 * self.X[:, 2] + 0.1 * self.X[:, 3] + 
                  0.05 * np.random.randn(self.n_samples)).astype(np.float64)
        self.X_test = np.random.randn(30, self.n_features).astype(np.float64)
        
    def test_basic_functionality(self):
        """Test basic fit and predict functionality"""
        import auroraml.adaboost as aml_adaboost
        
        model = aml_adaboost.AdaBoostRegressor(n_estimators=50, learning_rate=1.0, random_state=42)
        model.fit(self.X, self.y)
        predictions = model.predict(self.X_test)
        
        self.assertEqual(len(predictions), len(self.X_test))
        self.assertIsInstance(predictions, np.ndarray)
        self.assertFalse(np.any(np.isnan(predictions)))
        self.assertFalse(np.any(np.isinf(predictions)))
        
    def test_parameters(self):
        """Test parameter getter and setter"""
        import auroraml.adaboost as aml_adaboost
        
        model = aml_adaboost.AdaBoostRegressor(n_estimators=50, learning_rate=1.0, 
                                               loss='linear', random_state=42)
        
        params = model.get_params()
        self.assertIn('n_estimators', params)
        self.assertIn('learning_rate', params)
        self.assertIn('loss', params)
        self.assertIn('random_state', params)
        
        # Test parameter setting
        model.set_params(n_estimators=75, loss='square')
        updated_params = model.get_params()
        self.assertEqual(updated_params['n_estimators'], "75")
        self.assertEqual(updated_params['loss'], "square")
        
    def test_performance(self):
        """Test model performance"""
        import auroraml.adaboost as aml_adaboost
        import auroraml.metrics as aml_metrics
        
        model = aml_adaboost.AdaBoostRegressor(n_estimators=100, learning_rate=1.0, random_state=42)
        model.fit(self.X, self.y)
        predictions = model.predict(self.X)
        
        mse = aml_metrics.mean_squared_error(self.y, predictions)
        self.assertLess(mse, 10.0)
        
        r2 = aml_metrics.r2_score(self.y, predictions)
        self.assertGreater(r2, 0.5)
        
    def test_different_losses(self):
        """Test with different loss functions"""
        import auroraml.adaboost as aml_adaboost
        
        losses = ['linear', 'square', 'exponential']
        for loss in losses:
            model = aml_adaboost.AdaBoostRegressor(n_estimators=50, learning_rate=1.0, 
                                                  loss=loss, random_state=42)
            model.fit(self.X, self.y)
            predictions = model.predict(self.X_test)
            self.assertEqual(len(predictions), len(self.X_test))
            self.assertFalse(np.any(np.isnan(predictions)))
            
    def test_different_learning_rates(self):
        """Test with different learning rates"""
        import auroraml.adaboost as aml_adaboost
        
        learning_rates = [0.5, 1.0, 2.0]
        for lr in learning_rates:
            model = aml_adaboost.AdaBoostRegressor(n_estimators=50, learning_rate=lr, random_state=42)
            model.fit(self.X, self.y)
            predictions = model.predict(self.X_test)
            self.assertEqual(len(predictions), len(self.X_test))
            
    def test_is_fitted(self):
        """Test is_fitted method"""
        import auroraml.adaboost as aml_adaboost
        
        model = aml_adaboost.AdaBoostRegressor()
        self.assertFalse(model.is_fitted())
        
        model.fit(self.X, self.y)
        self.assertTrue(model.is_fitted())
        
    def test_consistency(self):
        """Test model consistency with same random seed"""
        import auroraml.adaboost as aml_adaboost
        
        model1 = aml_adaboost.AdaBoostRegressor(n_estimators=50, learning_rate=1.0, random_state=42)
        model2 = aml_adaboost.AdaBoostRegressor(n_estimators=50, learning_rate=1.0, random_state=42)
        
        model1.fit(self.X, self.y)
        model2.fit(self.X, self.y)
        
        pred1 = model1.predict(self.X_test)
        pred2 = model2.predict(self.X_test)
        
        np.testing.assert_allclose(pred1, pred2, rtol=1e-5)

if __name__ == '__main__':
    unittest.main()

File: tests/test_adaboost.cpp
#include <gtest/gtest.h>
#include "auroraml/adaboost.hpp"
#include "auroraml/metrics.hpp"
#include <Eigen/Dense>
#include <algorithm>
#include <vector>

namespace auroraml {
namespace test {

class AdaBoostTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Generate synthetic data
        n_samples = 200;
        n_features = 4;
        
        // Create regression data
        X = MatrixXd::Random(n_samples, n_features);
        y = VectorXd::Zero(n_samples);
        for (int i = 0; i < n_samples; ++i) {
            y(i) = 2.0 * X(i, 0) + 1.5 * X(i, 1) - 0.8 * X(i, 2) + 0.1 * X(i, 3) + 0.05 * (MatrixXd::Random(1, 1))(0, 0);
        }
        
        // Create classification data
        y_classification = VectorXd::Zero(n_samples);
        for (int i = 0; i < n_samples; ++i) {
            double score = X(i, 0) + X(i, 1) - 0.5 * X(i, 2);
            y_classification(i) = (score > 0.0) ? 1.0 : 0.0;
        }
        
        // Create multiclass data
        y_multiclass = VectorXd::Zero(n_samples);
        for (int i = 0; i < n_samples; ++i) {
            double score = X(i, 0) + X(i, 1);
            if (score > 1.0) {
                y_multiclass(i) = 2.0;
            } else if (score > -1.0) {
                y_multiclass(i) = 1.0;
            } else {
                y_multiclass(i) = 0.0;
            }
        }
        
        // Create test data
        X_test = MatrixXd::Random(30, n_features);
    }
    
    int n_samples, n_features;
    MatrixXd X, X_test;
    VectorXd y, y_classification, y_multiclass;
};

// AdaBoost Classifier Tests
TEST_F(AdaBoostTest, AdaBoostClassifierFit) {
    ensemble::AdaBoostClassifier adaboost(50, 1.0, 42);
    adaboost.fit(X, y_classification);
    
    EXPECT_TRUE(adaboost.is_fitted());
    
    std::vector<int> classes = adaboost.classes();
    EXPECT_GE(classes.size(), 2);
}

TEST_F(AdaBoostTest, AdaBoostClassifierPredictClasses) {
    ensemble::AdaBoostClassifier adaboost(50, 1.0, 42);
    adaboost.fit(X, y_classification);
    
    VectorXi y_pred = adaboost.predict_classes(X_test);
    EXPECT_EQ(y_pred.size(), X_test.rows());
    
    // Check that predictions are valid class labels
    std::vector<int> classes = adaboost.classes();
    for (int i = 0; i < y_pred.size(); ++i) {
        EXPECT_TRUE(std::find(classes.begin(), classes.end(), y_pred(i)) != classes.end());
    }
}

TEST_F(AdaBoostTest, AdaBoostClassifierPredictProba) {
    ensemble::AdaBoostClassifier adaboost(50, 1.0, 42);
    adaboost.fit(X, y_classification);
    
    MatrixXd y_proba = adaboost.predict_proba(X_test);
    EXPECT_EQ(y_proba.rows(), X_test.rows());
    EXPECT_EQ(y_proba.cols(), 2);  // Binary classification
    
    // Check that probabilities sum to 1
    for (int i = 0; i < y_proba.rows(); ++i) {
        double sum = y_proba.row(i).sum();
        EXPECT_NEAR(sum, 1.0, 1e-6);
    }
    
    // Check that probabilities are non-negative
    EXPECT_TRUE((y_proba.array() >= -1e-6).all());
    EXPECT_TRUE((y_proba.array() <= 1.0 + 1e-6).all());
}

TEST_F(AdaBoostTest, AdaBoostClassifierDecisionFunction) {
    ensemble::AdaBoostClassifier adaboost(50, 1.0, 42);
    adaboost.fit(X, y_classification);
    
    VectorXd decision = adaboost.decision_function(X_test);
    EXPECT_EQ(decision.size(), X_test.rows());
}

TEST_F(AdaBoostTest, AdaBoostClassifierParameters) {
    ensemble::AdaBoostClassifier adaboost(100, 0.5, 42);
    
    Params params = adaboost.get_params();
    EXPECT_EQ(params.at("n_estimators"), "100");
    EXPECT_EQ(params.at("learning_rate"), "0.500000");
    EXPECT_EQ(params.at("random_state"), "42");
    
    // Test set_params
    Params new_params = {{"n_estimators", "75"}, {"learning_rate", "0.8"}, {"random_state", "123"}};
    adaboost.set_params(new_params);
    Params updated = adaboost.get_params();
    EXPECT_EQ(updated.at("n_estimators"), "75");
}

TEST_F(AdaBoostTest, AdaBoostClassifierPerformance) {
    ensemble::AdaBoostClassifier adaboost(100, 1.0, 42);
    adaboost.fit(X, y_classification);
    
    VectorXi y_pred = adaboost.predict_classes(X);
    VectorXi y_true = y_classification.cast<int>();
    
    double accuracy = metrics::accuracy_score(y_true, y_pred);
    EXPECT_GT(accuracy, 0.7);  // Should achieve reasonable accuracy on training data
}

TEST_F(AdaBoostTest, AdaBoostClassifierDifferentLearningRates) {
    std::vector<double> learning_rates = {0.5, 1.0, 2.0};
    
    for (double lr : learning_rates) {
        ensemble::AdaBoostClassifier adaboost(50, lr, 42);
        adaboost.fit(X, y_classification);
        EXPECT_TRUE(adaboost.is_fitted());
        
        VectorXi y_pred = adaboost.predict_classes(X_test);
        EXPECT_EQ(y_pred.size(), X_test.rows());
    }
}

TEST_F(AdaBoostTest, AdaBoostClassifierDifferentNEstimators) {
    std::vector<int> n_estimators = {10, 50, 100};
    
    for (int n_est : n_estimators) {
        ensemble::AdaBoostClassifier adaboost(n_est, 1.0, 42);
        adaboost.fit(X, y_classification);
        EXPECT_TRUE(adaboost.is_fitted());
        
        VectorXi y_pred = adaboost.predict_classes(X_test);
        EXPECT_EQ(y_pred.size(), X_test.rows());
    }
}

// AdaBoost Regressor Tests
TEST_F(AdaBoostTest, AdaBoostRegressorFit) {
    ensemble::AdaBoostRegressor adaboost(50, 1.0, "linear", 42);
    adaboost.fit(X, y);
    
    EXPECT_TRUE(adaboost.is_fitted());
}

TEST_F(AdaBoostTest, AdaBoostRegressorPredict) {
    ensemble::AdaBoostRegressor adaboost(50, 1.0, "linear", 42);
    adaboost.fit(X, y);
    
    VectorXd y_pred = adaboost.predict(X_test);
    EXPECT_EQ(y_pred.size(), X_test.rows());
    
    // Predictions should be reasonable (not NaN or Inf)
    EXPECT_FALSE(y_pred.array().isNaN().any());
    EXPECT_FALSE(y_pred.array().isInf().any());
}

TEST_F(AdaBoostTest, AdaBoostRegressorPerformance) {
    ensemble::AdaBoostRegressor adaboost(100, 1.0, "linear", 42);
    adaboost.fit(X, y);
    
    VectorXd y_pred = adaboost.predict(X);
    
    double mse = metrics::mean_squared_error(y, y_pred);
    EXPECT_LT(mse, 10.0);  // Should achieve reasonable MSE on training data
    
    double r2 = metrics::r2_score(y, y_pred);
    EXPECT_GT(r2, 0.5);  // Should achieve reasonable R²
}

TEST_F(AdaBoostTest, AdaBoostRegressorDifferentLosses) {
    std::vector<std::string> losses = {"linear", "square", "exponential"};
    
    for (const std::string& loss : losses) {
        ensemble::AdaBoostRegressor adaboost(50, 1.0, loss, 42);
        adaboost.fit(X, y);
        EXPECT_TRUE(adaboost.is_fitted());
        
        VectorXd y_pred = adaboost.predict(X_test);
        EXPECT_EQ(y_pred.size(), X_test.rows());
    }
}

TEST_F(AdaBoostTest, AdaBoostRegressorParameters) {
    ensemble::AdaBoostRegressor adaboost(100, 0.5, "linear", 42);
    
    Params params = adaboost.get_params();
    EXPECT_EQ(params.at("n_estimators"), "100");
    EXPECT_EQ(params.at("learning_rate"), "0.500000");
    EXPECT_EQ(params.at("loss"), "linear");
    EXPECT_EQ(params.at("random_state"), "42");
    
    // Test set_params
    Params new_params = {{"n_estimators", "75"}, {"learning_rate", "0.8"}, {"loss", "square"}, {"random_state", "123"}};
    adaboost.set_params(new_params);
    Params updated = adaboost.get_params();
    EXPECT_EQ(updated.at("n_estimators"), "75");
    EXPECT_EQ(updated.at("loss"), "square");
}

TEST_F(AdaBoostTest, AdaBoostRegressorDifferentLearningRates) {
    std::vector<double> learning_rates = {0.5, 1.0, 2.0};
    
    for (double lr : learning_rates) {
        ensemble::AdaBoostRegressor adaboost(50, lr, "linear", 42);
        adaboost.fit(X, y);
        EXPECT_TRUE(adaboost.is_fitted());
        
        VectorXd y_pred = adaboost.predict(X_test);
        EXPECT_EQ(y_pred.size(), X_test.rows());
    }
}

TEST_F(AdaBoostTest, AdaBoostRegressorConsistency) {
    ensemble::AdaBoostRegressor adaboost1(50, 1.0, "linear", 42);
    ensemble::AdaBoostRegressor adaboost2(50, 1.0, "linear", 42);
    
    adaboost1.fit(X, y);
    adaboost2.fit(X, y);
    
    VectorXd pred1 = adaboost1.predict(X_test);
    VectorXd pred2 = adaboost2.predict(X_test);
    
    // With same random seed, predictions should be similar (may have small numerical differences)
    for (int i = 0; i < pred1.size(); ++i) {
        EXPECT_NEAR(pred1(i), pred2(i), 1e-5);
    }
}

} // namespace test
} // namespace auroraml

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

